From 9dcb6f4d5fb3d9a455996200a12ac1db1c7b3ea6 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Thu, 10 Oct 2024 09:43:57 +0200
Subject: [PATCH 0894/1013] evl/net: dev, input: fix unsafe bitwise operations
 on RX

The physical interface might be receiving packets while we are busy
setting/clearing the eBPF filter, causing bit loss since those
concurrent portions of the RX path are not serialized.

Always use atomic bitops on both sides when flipping bits in the
device state flags to fix this.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 kernel/evl/net/dev.c   |  4 ++--
 kernel/evl/net/input.c | 13 ++++++++-----
 2 files changed, 10 insertions(+), 7 deletions(-)

diff --git a/kernel/evl/net/dev.c b/kernel/evl/net/dev.c
index d13f7306e595..8342fd884d53 100644
--- a/kernel/evl/net/dev.c
+++ b/kernel/evl/net/dev.c
@@ -498,9 +498,9 @@ static void __set_rx_filter(struct evl_netdev_state *est,
 				lockdep_is_held(&est->filter_lock));
 	rcu_assign_pointer(est->rx_filter, filter);
 	if (filter)
-		__set_bit(EVL_NETDEV_RXFILTER_BIT, &est->flags);
+		set_bit(EVL_NETDEV_RXFILTER_BIT, &est->flags);
 	else
-		__clear_bit(EVL_NETDEV_RXFILTER_BIT, &est->flags);
+		clear_bit(EVL_NETDEV_RXFILTER_BIT, &est->flags);
 	spin_unlock_bh(&est->filter_lock);
 
 	if (old)
diff --git a/kernel/evl/net/input.c b/kernel/evl/net/input.c
index d9cf22866ceb..9ea9469f0561 100644
--- a/kernel/evl/net/input.c
+++ b/kernel/evl/net/input.c
@@ -35,9 +35,10 @@ static void napi_poll_oob(struct evl_netdev_state *est) /* oob */
 
 	/*
 	 * We are about to drop the RX lock, clear this flag early to
-	 * close a race.
+	 * close a race. We might compete with __set_rx_filter(), so
+	 * use atomic bitops.
 	 */
-	__clear_bit(EVL_NETDEV_POLL_SCHED, &est->flags);
+	clear_bit(EVL_NETDEV_POLL_SCHED, &est->flags);
 
 	list_for_each_entry_safe(napi, tmp, &est->rx_poll, poll_list) {
 		int budget = napi->weight;
@@ -62,7 +63,7 @@ static void napi_poll_oob(struct evl_netdev_state *est) /* oob */
 
 	if (!list_empty(&requeuing)) {
 		list_splice(&requeuing, &est->rx_poll);
-		__set_bit(EVL_NETDEV_POLL_SCHED, &est->flags);
+		set_bit(EVL_NETDEV_POLL_SCHED, &est->flags);
 	}
 
 	raw_spin_unlock_irqrestore(&est->rx_lock, flags);
@@ -206,11 +207,13 @@ void napi_schedule_oob(struct napi_struct *n) /* oob */
 	/*
 	 * We might have multiple NAPI instances per device, so
 	 * serialization is required despite a single NAPI instance
-	 * may be active at any point in time. Oh, well.
+	 * may be active at any point in time. Oh, well. See
+	 * napi_poll_oob() for an explanation about the requirement
+	 * for atomic bitops (EVL_NETDEV_POLL_SCHED).
 	 */
 	raw_spin_lock_irqsave(&est->rx_lock, flags);
 	list_add(&n->poll_list, &est->rx_poll);
-	__set_bit(EVL_NETDEV_POLL_SCHED, &est->flags);
+	set_bit(EVL_NETDEV_POLL_SCHED, &est->flags);
 	raw_spin_unlock_irqrestore(&est->rx_lock, flags);
 	evl_net_wake_rx(dev);
 }
-- 
2.30.2

